query-planner-js/src/Scope.ts

```diff
@@ -0,0 +1,271 @@
+      // so it's not useful anymore. Skip it, and check if we can prune further.
+      return Scope.pruneRefinedTypes(toPrune.enclosing, refiningType);
+    }
+    return new Scope(
+      toPrune.context,
+      toPrune.parentType,
+      toPrune.directives,
+      Scope.pruneRefinedTypes(toPrune.enclosing, refiningType)
+    );
+  }
+
+  /**
+   * Whether refining this scope by the provided type would strictly reduce the possible runtime types
+   * of the scope.
+   */
+  isStrictlyRefinedBy(type: GraphQLCompositeType) : boolean {
+    // If one of the type in the chain is a subtype of the provided type, then the type will not strictly
+    // refine the scope.
+    let scope: Scope | undefined = this;
+    while (scope) {
+      if (isTypeSubTypeOf(this.context.schema, scope.parentType, type)) {
+        return false;
+      }
+      scope = scope.enclosing;
+    }
+    return true;
+  }
+
+  /**
+   *  Whether this scope is restricting the possible runtime types of the provided type.
+   */
+  isStrictlyRefining(type: GraphQLCompositeType) : boolean {
+    // This scope will refine the provided type, unless that provided type is a subtype of all
+    // the type in the chain.
+    let scope: Scope | undefined = this;
+    while (scope) {
+      if (scope.parentType !== type && isTypeSubTypeOf(this.context.schema, scope.parentType, type)) {
+        return true;
+      }
+      scope = scope.enclosing;
+    }
+    return false;
+  }
+
+  private computePossibleRuntimeTypes() : ReadonlyArray<GraphQLObjectType> {
+    // The possible runtime types is the intersection of all the possible types of each condition in scope.
+    let possibleTypes = this.context.getPossibleTypes(this.parentType);
+    let nextScope = this.enclosing;
+    while (nextScope) {
+      let enclosingPossibleTypes = this.context.getPossibleTypes(nextScope.parentType);
+      possibleTypes = possibleTypes.filter(t => enclosingPossibleTypes.includes(t));
+      nextScope = nextScope.enclosing;
+    }
+    return possibleTypes;
+  }
+
+  /**
+   * Computes the set of object types that an object within this scope can have at runtime.
+   */
+  possibleRuntimeTypes() : ReadonlyArray<GraphQLObjectType> {
+    if (!this.cachedRuntimeTypes) {
+      this.cachedRuntimeTypes = this.computePossibleRuntimeTypes();
+    }
+    return this.cachedRuntimeTypes;
+  }
+
+  private static valueIdentityKey(value: ValueNode) : string {
+    // We distinguish different value types though a leading character with a comma (so the int 3 has key "i'3") to
+    // avoid conflicts.
+    switch (value.kind) {
+      case 'Variable':
+        return value.name.value;
+      case 'IntValue':
+        return "i'" + value.value;
+      case 'FloatValue':
+        return "f'" + value.value;
+      case 'EnumValue':
+        return "e'" + value.value;
+      case 'StringValue':
+        return "s'" + value.value;
```

Comment:
If it's important that "same IK implies equivalent" then this encoding combined with the list and object encodings is not strong enough to achieve that goal: there's a conflict between `["x", "y-s'z"]` and `["x-s'y", "z"]`, for example.

I find in TS that the best way to not have to think hard about string escaping is to just use JSON.stringify.  I think you could make this line be

```
return `s'${ JSON.stringify(value.value) }`;
```

so that the encoding of a string ends up being `s'"foo"` (with those double quotes). I think that's sufficient to make things unambiguous (assuming that all the values for enum name, variable name, field name, etc are valid GraphQL identifiers).

Alternatively, if the only thing keeping you away from using the GraphQL printer is field ordering, we do have [a `sortAST` function in our `apollo-graphql` package](https://github.com/apollographql/apollo-tooling/blob/9b8f34ca0aeb7ec1355eb4af6bdbc47802935609/packages/apollo-graphql/src/transforms.ts#L102-L159) which is already a dependency of our gateway which can be used for this sort of normalization.  (Although it perhaps needs to be updated for the more recent version of the graphql spec where directives can be repeatable, as it assumes you can sort directive applications just by directive name.)

Reply:
Good point about strings in list. I've switched to `JSON.stringify` and fixed the comments typos/errors.